core/src/main/java/hudson/model/labels/LabelExpression.java

```diff
@@ -210,4 +225,156 @@ public LabelOperatorPrecedence precedence() {
             return LabelOperatorPrecedence.IMPLIES;
         }
     }
+
+    //region Auto-Completion and Validation
+
+    /**
+     * Utility class for taking the current input value and computing a list of potential terms to match against the
+     * list of defined labels.
+     */
+    static class AutoCompleteSeeder {
+        private String source;
+
+        AutoCompleteSeeder(String source) {
+            this.source = source;
+        }
+
+        List<String> getSeeds() {
+            ArrayList<String> terms = new ArrayList<>();
+            boolean trailingQuote = source.endsWith("\"");
+            boolean leadingQuote = source.startsWith("\"");
+            boolean trailingSpace = source.endsWith(" ");
+
+            if (trailingQuote || (trailingSpace && !leadingQuote)) {
+                terms.add("");
+            } else {
+                if (leadingQuote) {
+                    int quote = source.lastIndexOf('"');
+                    if (quote == 0) {
+                        terms.add(source.substring(1));
+                    } else {
+                        terms.add("");
+                    }
+                } else {
+                    int space = source.lastIndexOf(' ');
+                    if (space > -1) {
+                        terms.add(source.substring(space+1));
+                    } else {
+                        terms.add(source);
+                    }
+                }
+            }
+
+            return terms;
+        }
+    }
+
+    /**
+     * Plugins may want to contribute additional restrictions on the use of specific labels for specific context items.
+     * This extension point allows such restrictions.
+     *
+     * @since TODO
+     */
+    public interface LabelValidator extends ExtensionPoint {
+
+        /**
+         * Validates the use of a label within a particular context.
+         *
+         * @param item  The context item to be restricted by the label.
+         * @param label The label that the job wants to restrict itself to.
+         * @return The validation result.
+         */
+        @NonNull
+        FormValidation check(@NonNull Item item, @NonNull Label label);
+
+    }
+
+    /**
+     * Generates auto-completion candidates for a (partial) label.
+     *
+     * @param label The (partial) label for which auto-completion is being requested.
+     * @return A set of auto-completion candidates.
+     * @since TODO
+     */
+    public static AutoCompletionCandidates autoComplete(@Nullable String label) {
```

Comment:
```suggestion
    @NonNull
    public static AutoCompletionCandidates autoComplete(@Nullable String label) {
```

Reply:
Yes. After applying the other suggested changes, I'll make another pass to ensure all new/changed API has complete nullability annotations.